dnf5/shared_options.cpp

```diff
@@ -115,11 +115,6 @@ void create_offline_option(dnf5::Command & command) {
                                      [[maybe_unused]] libdnf5::cli::ArgumentParser::NamedArg * arg,
                                      [[maybe_unused]] const char * option,
                                      [[maybe_unused]] const char * value) {
-        // The installroot will be prepended to the cachedir and system_cachedir later in Base.setup()
-        const auto & offline_datadir = dnf5::offline::DEFAULT_DATADIR;
-        ctx.get_base().get_config().get_system_cachedir_option().set(
-            libdnf5::Option::Priority::INSTALLROOT, offline_datadir);
-        ctx.get_base().get_config().get_cachedir_option().set(libdnf5::Option::Priority::INSTALLROOT, offline_datadir);
```

Comment:
> I might be missing something but I don't see how [5f8b76a](https://github.com/rpm-software-management/dnf5/commit/5f8b76ad7e0faa5c5b151e4301417f004bc4470e) broke system upgrade. I think that the problem was rather this code from `shared_options.cpp` because the `system-upgrade` command doesn't call it (it doesn't have `--offline` option) so the `cachedir` wasn't set correctly and when `offline _execute` was looking for it it wasn't there.

I think the `cachedir`/`system_cachedir` being incorrect was a separate issue. Yes, system-upgrade was using `/var/cache/libdnf5`, but it still worked. `offline _execute` reads the `cachedir`/`system_cachedir` from `offline-transaction-state.toml`, which was being set to `/var/cache/libdnf5`.

I think the bigger issue is that since https://github.com/rpm-software-management/dnf5/pull/1264, the paths to RPMs aren't stored in the transaction.json unless a `packages_path` is passed to `Transaction::serialize`, which `store_offline` was not doing. For some reason, simple `system-upgrade download` transactions, such as a single package upgrade, still worked, but when I tried upgrading from F39 to F40, I saw these errors: https://github.com/rpm-software-management/dnf5/issues/1446. To check, I just tried upgrading F39 to F40 on dd2c73581655f7a491b2f0c20f14718298c02ea3, the last commit before #1264, and it worked.

> 
> Anyway this is fixed with your PR.
> 
> I was also confused why do all CI tests pass because I think we test this. I have quickly looked into it and there is probably a problem in the way we run the tests. For example the `_execute` command is run like: `dnf5 -y --installroot=/tmp/dnf_ci_installroot_l9slxa11 --releasever=30 --setopt=module_platform_id=platform:f29 --setopt=cachedir=/tmp/dnf_ci_installroot_l9slxa11/var/cache/dnf offline _execute` but the step says only: `When I execute dnf with args "offline _execute"`. I fear that since the CI is adding both `releasever` and `cachedir` by itself it can skew the results.

That shouldn't be a problem now since `system-upgrade` and `--offline` will not change the `cachedir`. But I think the reason the tests were passing has more to do with the tests having too simple transactions (maybe something to do with comps/groups?)

Reply:
> That shouldn't be a problem now since system-upgrade and --offline will not change the cachedir. But I think the reason the tests were passing has more to do with the tests having too simple transactions (maybe something to do with comps/groups?)

For the record, the issue was that `offline _execute` did not load the `target_releasever` from `offline-transaction-state.toml`, and the repositories for the new releasever were not being loaded. That doesn't matter anymore since now we aren't loading repositories at all during `offline _execute`.